File: moped-etl/ecapris-statuses/process/time.py
from datetime import datetime, timezone
from zoneinfo import ZoneInfo


def convert_to_timezone_aware_timestamp(date_str):
    """
    Converts a timestamp string in the format 'YYYY-MM-DD HH:MM:SS' to a timezone-aware
    timestamp in the 'America/Chicago' timezone.

    Args:
        date_str (str): The timestamp string to convert, in the format 'YYYY-MM-DD HH:MM:SS'.

    Returns:
        str: The ISO timestamp string in the 'America/Chicago' timezone.
    """
    # Update the date format to match your input without milliseconds
    date_format = "%Y-%m-%d %H:%M:%S"

    # Parse the timestamp string
    parsed_date = datetime.strptime(date_str, date_format)

    # Convert to timezone-aware by attaching the America/Chicago timezone
    parsed_date_tz = parsed_date.replace(tzinfo=ZoneInfo("America/Chicago"))

    # Convert to ISO format
    timestamptz = parsed_date_tz.isoformat()

    return timestamptz

File: toolbox/backfill_phb_completions/helpers/time.ts
/**
 * Socrata returns a floating timestamp without timezone info and date only. So, we need to update to be
 * midnight central time and then convert to an ISO string for Hasura to store as a timestamptz.
 * See https://dev.socrata.com/docs/datatypes/floating_timestamp.html
 * Example input: "2023-01-15T00:00:00.000"
 * Example output: "2023-01-15T06:00:00.000Z" or "2023-01-15T05:00:00.000Z" depending on daylight savings time.
 */
export function toTimestamptz(floatingTimestamp: string): string {
  const dateStr = floatingTimestamp.split("T")[0];
  const centralMidnight = new Date(
    new Date(`${dateStr}T00:00:00`).toLocaleString("en-US", {
      timeZone: "America/Chicago",
    }),
  );
  return centralMidnight.toISOString();
}
